Userland/Services/DeviceMapper/DeviceEventLoop.cpp

```diff
@@ -0,0 +1,234 @@
+    { {}, "render"sv, "gpu/render%d"sv, false, 28, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0666 },
+    { "window"sv, "gpu-connector"sv, "gpu/connector%d"sv, false, 226, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0660 },
+    { {}, "virtio-console"sv, "hvc0p%d"sv, false, 229, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0666 },
+    { "phys"sv, "hid-mouse"sv, "input/mouse/%d"sv, false, 10, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0666 },
+    { "phys"sv, "hid-keyboard"sv, "input/keyboard/%d"sv, false, 85, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0666 },
+    { {}, "storage"sv, "hd%c"sv, true, 3, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0600 },
+    { "tty"sv, "console"sv, "tty%d"sv, false, 4, MinorNumberAllocationType::SequentialLimited, 0, 63, 0620 },
+    { "tty"sv, "console"sv, "ttyS%d"sv, false, 4, MinorNumberAllocationType::SequentialLimited, 64, 127, 0620 },
+};
+
+static bool is_in_minor_number_range(DeviceEventLoop::device_node_match const& matcher, MinorNumber minor_number)
+{
+    if (matcher.minor_number_allocation_type == MinorNumberAllocationType::SequentialUnlimited)
+        return true;
+
+    return matcher.minor_number_start <= minor_number && static_cast<MinorNumber>(matcher.minor_number_start.value() + matcher.minor_number_range_size) >= minor_number;
+}
+
+static DeviceEventLoop::device_node_match const* device_node_family_to_match_type(bool block_device, MajorNumber major_number, MinorNumber minor_number)
+{
+    for (auto& matcher : s_matchers) {
+        if (matcher.major_number == major_number
+            && block_device == matcher.block_device
+            && is_in_minor_number_range(matcher, minor_number))
+            return &matcher;
+    }
+    return nullptr;
+}
+
+static bool is_in_family_minor_number_range(DeviceNodeFamily const& family, MinorNumber minor_number)
+{
+    return family.base_minor_number() <= minor_number && static_cast<MinorNumber>(family.base_minor_number().value() + family.devices_symbol_suffix_allocation_map().size()) >= minor_number;
+}
+
+DeviceNodeFamily* DeviceEventLoop::find_device_node_family(bool block_device, MajorNumber major_number, MinorNumber minor_number) const
+{
+    DeviceNodeFamily::Type type = block_device ? DeviceNodeFamily::Type::CharacterDevice : DeviceNodeFamily::Type::CharacterDevice;
+    for (auto& family : m_device_node_families) {
+        if (family->major_number() == major_number && type == family->type() && is_in_family_minor_number_range(*family, minor_number))
+            return family.ptr();
+    }
+    return nullptr;
+}
+
+ErrorOr<DeviceNodeFamily*> DeviceEventLoop::find_or_register_new_device_node_family(device_node_match const& match, bool block_device, MajorNumber major_number, MinorNumber minor_number)
+{
+    if (auto* family = find_device_node_family(block_device, major_number, minor_number); !!family)
```

Comment:
(Sidenote for using references instead of pointers: this will also become more readable; instead of the weird `!!` you'll get `family.has_value()`)